src/components/Tag.ts

```diff
@@ -0,0 +1,200 @@
+                disabled: this.props.readOnly,
+                inputProps,
+                inputValue: this.state.newTag,
+                onChangeInput: this.handleChangeInput,
+                maxTags: this.props.tagLimit === 0 ? undefined : this.props.tagLimit,
+                onChange: this.handleChange,
+                renderInput: this.props.enableSuggestions ? this.renderAutosuggest : undefined,
+                value: this.state.tagList
+            }),
+            createElement(Alert, {
+                bootstrapStyle: "danger",
+                className: "widget-tag-alert",
+                message: this.state.alertMessage
+            })
+        );
+    }
+
+    componentDidMount() {
+        const tagInputSelector = document.querySelectorAll(".react-tagsinput-input");
+        this.addEvents(tagInputSelector);
+    }
+
+    componentWillReceiveProps(newProps: TagProps) {
+        this.setState({
+            alertMessage: newProps.alertMessage,
+            tagList: newProps.tagList
+        });
+    }
+
+    componentWillUnmount() {
+        const inputNodeList = document.querySelectorAll(".react-tagsinput-input");
+
+        for (let i = 0; inputNodeList[i]; i++) {
+            inputNodeList[i].removeEventListener("focus");
+            inputNodeList[i].removeEventListener("blur");
+        }
+    }
+
+    private addEvents(nodes: NodeListOf<Element>) {
+        for (let i = 0; nodes[i]; i++) {
+            const node = nodes[i] as HTMLElement;
+
+            node.addEventListener("focus", this.hundleFocus);
+            node.addEventListener("blur", this.hundleOnblur);
+        }
+    }
+
+    private handleChangeInput(newTag: string) {
+        const { tagLimit, tagLimitMessage, showError } = this.props;
+
+        if (tagLimit === 0) {
+            this.setState({ newTag });
+        } else if (this.state.tagList.length >= tagLimit) {
+            showError(tagLimitMessage.replace("{limit}", `${tagLimit}`));
+            this.setState({ newTag });
+        } else {
+            this.setState({ newTag });
+        }
+    }
+
+    private renderAutosuggest() {
+        return createElement(AutoComplete, {
+            addTag: (tag: string) => this.processTag(tag),
+            fetchSuggestions: this.props.fetchSuggestions,
+            inputPlaceholder: this.props.inputPlaceholder,
+            readOnly: this.props.readOnly,
+            lazyLoad: this.props.lazyLoad,
+            suggestions: processSuggestions(this.props.suggestions)
+        });
+    }
+
+    private processTag(tag: string) {
+        const { tagLimit, tagLimitMessage, createTag } = this.props;
+        const tagList = this.state.tagList;
+
+        if (tagLimit === 0 || tagLimit + 1 > this.state.tagList.length) {
+            // Validate tag if its not a duplicate.
+            if (this.validateTagInput(tag, this.state.tagList)) {
+                this.setState ({
```

Comment:
inline, can be one line

Reply:
There are more than one properties in with the updated code..

Reply:
Done